storm-core/src/clj/backtype/storm/daemon/nimbus.clj

```diff
@@ -1767,6 +1900,169 @@
+          (let [ret (.read is)]
+            (.put downloaders session is)
+            (when (empty? ret)
+              (.close is)
+              (.remove downloaders session))
+            (log-debug "Sending " (alength ret) " bytes")
+            (ByteBuffer/wrap ret))))
+
+      (^void deleteBlob [this ^String blob-key]
+        (let [subject (->> (ReqContext/context)
+                           (.subject))]
+        (.deleteBlob (:blob-store nimbus) blob-key subject)
+        (when (instance? LocalFsBlobStore blob-store)
+          (.remove-blobstore-key! (:storm-cluster-state nimbus) blob-key)
+          (.remove-key-version! (:storm-cluster-state nimbus) blob-key))
+        (log-message "Deleted blob for key " blob-key)))
+
+      (^ListBlobsResult listBlobs [this ^String session]
+        (let [listers (:blob-listers nimbus)
+              ^Iterator keys-it (if (clojure.string/blank? session)
+                                  (->> (ReqContext/context)
+                                    (.subject)
+                                    (.listKeys (:blob-store nimbus)))
+                                  (.get listers session))
+              _ (or keys-it (throw-runtime "Blob list for session "
+                              session
+                              " does not exist (or timed out)"))
+
+              ;; Create a new session id if the user gave an empty session string.
+              ;; This is the use case when the user wishes to list blobs
+              ;; starting from the beginning.
+              session (if (clojure.string/blank? session)
+                        (let [new-session (uuid)]
+                          (log-message "Creating new session for downloading list " new-session)
+                          new-session)
+                        session)]
+          (if-not (.hasNext keys-it)
+            (do
+              (.remove listers session)
+              (log-message "No more blobs to list for session " session)
+              ;; A blank result communicates that there are no more blobs.
+              (ListBlobsResult. (ArrayList. 0) (str session)))
+            (let [^List list-chunk (->> keys-it
+                                     (iterator-seq)
+                                     (take 100) ;; Limit to next 100 keys
+                                     (ArrayList.))
+                  _ (log-message session " downloading " (.size list-chunk) " entries")]
+              (.put listers session keys-it)
+              (ListBlobsResult. list-chunk (str session))))))
+
+      (^int getBlobReplication [this ^String blob-key]
+        (->> (ReqContext/context)
+          (.subject)
+          (.getBlobReplication (:blob-store nimbus) blob-key)))
+
+      (^int updateBlobReplication [this ^String blob-key ^int replication]
+        (->> (ReqContext/context)
+          (.subject)
+          (.updateBlobReplication (:blob-store nimbus) blob-key replication)))
+      ;;Blobstore implementation code ends
```

Comment:
Remove this comment?  It should be clear whether a method is for the blobstore, and if we really want to separate the code, we could move these methods to their own namespace and `extend` the nimbus instance.